update command line documentation (#3117)
aiidateam/aiida-core: docs/source/verdi/verdi_user_guide.rst

                         soon as it's scheduled.
    -h, --help           Show this message and exit.

All help strings consist of three parts:

  * A ``Usage:`` line describing how to invoke the command
  * A description of the command's functionality
  * A list of the available options

The ``Usage:`` line encodes information on the command's parameters, e.g.:

 * ``[OPTIONS]``: this command takes one (or more) options

>  * ``PROCESSES``: this command requires a process as a positional argument

Reviewer: Since the importance here is the distinction of _requires_ and _optional_, maybe highlight these words?

 * ``[PROCESSES]``: this command takes a process as an optional positional argument
 * ``[PROCESSES]...``: this command takes one or more processes as (optional) positional arguments

Multi-value options are followed by ``...`` in the help string and the ``Usage:`` line

> of the corresponding command will contains the 'endopts' marker. For example::

Reviewer: Should be one line (or split at `[...] marker. For example::`)

  Usage: verdi export create [OPTIONS] [--] OUTPUT_FILE

    Export various entities, such as Codes, Computers, Groups and Nodes, to an
    archive file for backup or sharing purposes.

  Options:
    -X, --codes CODE...             one or multiple codes identified by their
                                    ID, UUID or label
    -Y, --computers COMPUTER...     one or multiple computers identified by
                                    their ID, UUID or label
    -G, --groups GROUP...           one or multiple groups identified by their
                                    ID, UUID or name
    -N, --nodes NODE...             one or multiple nodes identified by their ID

>                                     or UUID

Reviewer: There are more options here for `verdi export create`, could you show this perhaps by having a line like "`...`" or explicitly stating that this is a part of the full help response?

.. _cli_profile:

Profile
-------

> AiiDA supports multiple profiles per installation, one of which is marked as the default

Reviewer: Take care to create new lines "properly" :)

and used unless another profile is requested.
Show the current default profile using::

  verdi profile list

In order to use a different profile, pass the ``-p/--profile`` option to any ``verdi`` command, for example::

  verdi -p <profile> process list

Note that the specified profile will be used for this and *only* this command.
Use ``verdi profile setdefault`` in order to permanently change the default profile.

.. _cli_identifiers:

Identifiers
-----------

When working with AiiDA entities, you need a way to *refer* to them on the command line.
Any entity in AiiDA can be addressed via three identifiers:

>  * "Primary Key" (PK): An integer, e.g. ``723``, that identifies your entity within your database (automatically assigned)

Reviewer: This is great, however, could you afterwards (or here, but preferably afterwards) even more explicitly explain that PKs are local identifiers _only_ and the exact same database content may have completely different PKs for another user. And possibly add the now obvious advantage of UUIDs :)

Author: thanks - I've put in a sentence on this

 * `Universally Unique Identifier <https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)>`_ (UUID): A string, e.g. ``ce81c420-7751-48f6-af8e-eb7c6a30cec3`` that identifies your entity globally (automatically assigned)
 * Label: A human-readable string, e.g. ``test_calculation`` (manually assigned)

> Any ``verdi`` command that expects and identifier as a paramter will accept PKs, UUIDs and labels.

Reviewer: ```suggestion
Any ``verdi`` command that expects an identifier as a parameter will accept PKs, UUIDs, and labels.
```

                         soon as it's scheduled.
    -h, --help           Show this message and exit.

All help strings consist of three parts:

  * A ``Usage:`` line describing how to invoke the command
  * A description of the command's functionality
  * A list of the available options

The ``Usage:`` line encodes information on the command's parameters, e.g.:

 * ``[OPTIONS]``: this command takes one (or more) options
 * ``PROCESSES``: this command *requires* a process as a positional argument
 * ``[PROCESSES]``: this command takes a process as an *optional* positional argument
 * ``[PROCESSES]...``: this command takes one or more processes as *optional* positional arguments

> Multi-value options are followed by ``...`` in the help string and the ``Usage:`` line of the corresponding command will contains the 'endopts' marker. For example::

Reviewer: ```suggestion
Multi-value options are followed by ``...`` in the help string and the ``Usage:`` line of the corresponding command will contain the 'endopts' marker. For example::
```

  Usage: verdi export create [OPTIONS] [--] OUTPUT_FILE

    Export various entities, such as Codes, Computers, Groups and Nodes, to an
    archive file for backup or sharing purposes.

  Options:
    -X, --codes CODE...             one or multiple codes identified by their
                                    ID, UUID or label
    -Y, --computers COMPUTER...     one or multiple computers identified by
                                    their ID, UUID or label
    -G, --groups GROUP...           one or multiple groups identified by their
                                    ID, UUID or name
    -N, --nodes NODE...             one or multiple nodes identified by their ID
                                    or UUID
    ...

.. _cli_profile:

Profile
-------

> AiiDA supports multiple profiles per installation, one of which is marked as the default and used unless another profile is requested

Reviewer: ```suggestion
AiiDA supports multiple profiles per installation, one of which is marked as the default and used unless another profile is requested.
```

Show the current default profile using::

  verdi profile list

In order to use a different profile, pass the ``-p/--profile`` option to any ``verdi`` command, for example::

  verdi -p <profile> process list

Note that the specified profile will be used for this and *only* this command.
Use ``verdi profile setdefault`` in order to permanently change the default profile.

.. _cli_identifiers:

Identifiers
-----------

When working with AiiDA entities, you need a way to *refer* to them on the command line.
Any entity in AiiDA can be addressed via three identifiers:

>  * "Primary Key" (PK): An integer, e.g. ``723``, that identifies your entity within your database (automatically assigned)

Reviewer: ```suggestion
 * "Primary Key" (PK): An integer, e.g. ``723``, identifying your entity within your database (automatically assigned)
```

 * `Universally Unique Identifier <https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)>`_ (UUID): A string, e.g. ``ce81c420-7751-48f6-af8e-eb7c6a30cec3`` that identifies your entity globally (automatically assigned)
 * Label: A human-readable string, e.g. ``test_calculation`` (manually assigned)

.. note::

   PKs are easy to type and work as long as you stay within your database.

>    **When sharing data with others, however, always use UUIDs.**

Reviewer: Excellent! Thanks :)

Any ``verdi`` command that expects an identifier as a paramter will accept PKs, UUIDs and labels.

In almost all cases, this will work out of the box.
Since command line parameters are passed as strings, AiiDA needs to deduce the type of identifier from its content, which can fail in edge cases (see :ref:`cli_identifier_resolution` for details).
You can take the following precautions in order to avoid such edge cases:

  * PK: no precautions needed

>   * UUID: no precautions needed for full UUIDs. Partial UUIDs should include at least one non-numeric character or dash.

Reviewer: ```suggestion
  * UUID: no precautions needed for full UUIDs. Partial UUIDs should include at least one non-numeric character or dash
```

You don't have an ending period in any of your other lists (as far as I remember).

  * Label: add an exclamation mark ``!`` at the end of the identifier in order to force interpretation as a label

.. _cli_identifier_resolution:

Implementation of identifier resolution
.......................................

The logic for deducing the identifier type is as follows:

 1. Try interpreting the identifier as a PK (integer)

>  2. If this fails, try interpreting the identifier as a UUID (full or partial).

Reviewer: ```suggestion
 2. If this fails, try interpreting the identifier as a UUID (full or partial)
```

Same as above.

 3. If this fails, interpret the identifier as a label

The following example illustrates edge cases that can arise in this logic:

===  =====================================  ========
PK   UUID                                   LABEL
===  =====================================  ========
10   12dfb104-7b2b-4bca-adc0-1e4fd4ffcc88   group
11   deadbeef-62ba-444f-976d-31d925dac557   10
12   3df34a1e-5215-4e1a-b626-7f75b9586ef5   deadbeef
===  =====================================  ========

>  * trying to identify the first entity by its partial UUID ``12`` would match the third entity by its PK instead.

Reviewer: ```suggestion
 * trying to identify the first entity by its partial UUID ``12`` would match the third entity by its PK instead
```

 * trying to identify the second entity by its label ``10`` would match the first entity by its PK instead

>  * trying to identify the third entity by its label ``deadbeef`` would match the second entity on its partial UUID ``deadbeef`` instead.

Reviewer: ```suggestion
 * trying to identify the third entity by its label ``deadbeef`` would match the second entity on its partial UUID ``deadbeef`` instead
```
